### PR TITLE
fixes for tkinter in nltk.app and nltk.draw packages

### PR DESCRIPTION
**Pull request by hleeldc**
In the nltk.app and nltk.draw packages, updated modules that use tkinter by running 2to3.py. Manually fixed some spots that 2to3.py missed.

To ensure backward compatibility with python2.6, updated
- nltk/compat.py to add an import hook (PEP 302) for tkinter
- nltk/util.py to update in_idle function which was broken for py3

### CODE REVIEW

**Comment by kmike** [comment]
Clever trick with import hook! But honestly, I think the less clever approach would be better. Something like this:

```
# compat.py
# ...
if PY3:
    import tkinter.font as tkinter_font
else:
    import tkFont as tkinter_font

# app code
from nltk import compat

compat.tkinter_font
```

Pros of a simpler approach:
- import nltk.compat doesn't have global side effects;
- less code;
- code is simpler.

Cons:
- instead of `tkinter.font` nltk.app code would have `compat.tkinter_font` so it wouldn't be possible to just remove "nltk.compat" and have Python3 code.

File: nltk/draw/dispersion.py [hunk 1]
```diff
@@ -41,7 +41,7 @@ def dispersion_plot(text, words, ignore_case=False):
                     for y in range(len(words_to_comp))
                     if text_to_comp[x] == words_to_comp[y]]
     if points:
-        x, y = zip(*points)
+        x, y = list(zip(*points))
```
Comment (kmike) [inline nltk/draw/dispersion.py:44]: list() is not necessary here

File: nltk/app/chartparser_app.py [hunk 2]
```diff
@@ -216,7 +217,7 @@ def remove_callback(self, event, func=None):
 
     def _fire_callbacks(self, event, *args):
         if event not in self._callbacks: return
-        for cb_func in self._callbacks[event].keys(): cb_func(*args)
+        for cb_func in list(self._callbacks[event].keys()): cb_func(*args)
```
Comment (kmike) [inline nltk/app/chartparser_app.py:220]: I didn't check the code deeply, but why is `for cb_func in self.callbacks[event]:` not enough?

File: nltk/app/chunkparser_app.py [hunk 3]
```diff
@@ -854,7 +855,7 @@ def show_help(self, tab):
         for (name, tabstops, text) in self.HELP:
             if name == tab:
                 text = text.replace('<<TAGSET>>', '\n'.join(
-                    ('\t%s\t%s' % item for item in sorted(self.tagset.items(),
+                    ('\t%s\t%s' % item for item in sorted(list(self.tagset.items()),
```
Comment (kmike) [inline nltk/app/chunkparser_app.py:858]: list() is not necessary here

File: nltk/app/collocations_app.py [hunk 4]
```diff
@@ -284,7 +285,7 @@ def load_corpus(self, name):
 
     def non_default_corpora(self):
         copy = []
-        copy.extend(self.CORPORA.keys())
+        copy.extend(list(self.CORPORA.keys()))
```
Comment (kmike) [inline nltk/app/collocations_app.py:288]: list() is not necessary here

File: nltk/draw/table.py [hunk 5]
```diff
@@ -365,7 +366,7 @@ def insert(self, index, *rows):
             if len(elt) != len(self._column_names):
                 raise ValueError('rows should be tuples whose length '
                                  'is equal to the number of columns')
-        for (lb,elts) in zip(self._listboxes, zip(*rows)):
+        for (lb,elts) in zip(self._listboxes, list(zip(*rows))):
```
Comment (kmike) [inline nltk/draw/table.py:369]: list() is not necessary here

File: nltk/draw/table.py [hunk 6]
```diff
@@ -298,8 +299,8 @@ def configure(self, cnf={}, **kw):
                 >>> mlb.configure(label_foreground='red')
                 >>> mlb.configure(listbox_foreground='red')
         """
-        cnf = dict(cnf.items() + kw.items())
-        for (key, val) in cnf.items():
+        cnf = dict(list(cnf.items()) + list(kw.items()))
+        for (key, val) in list(cnf.items()):
```
Comment (kmike) [inline nltk/draw/table.py:303]: list() is not necessary

File: nltk/draw/tree.py [hunk 7]
```diff
@@ -430,7 +431,7 @@ def tree_to_treesegment(canvas, t, make_node=TextWidget,
     leaf_attribs = {}
     loc_attribs = {}
 
-    for (key, value) in attribs.items():
+    for (key, value) in list(attribs.items()):
```
Comment (kmike) [inline nltk/draw/tree.py:434]: list() is not necessary

File: nltk/draw/tree.py [hunk 8]
```diff
@@ -544,18 +545,18 @@ def bind_click_trees(self, callback, button=1):
...
     def bind_drag_trees(self, callback, button=1):
         """
         Add a binding to all tree segments.
         """
-        for tseg in self._expanded_trees.values():
+        for tseg in list(self._expanded_trees.values()):
             tseg.bind_drag(callback, button)
-        for tseg in self._collapsed_trees.values():
+        for tseg in list(self._collapsed_trees.values()):
             tseg.bind_drag(callback, button)
```
Comment (kmike) [inline nltk/draw/tree.py:561]: list() may be removed

**Comment by kmike** [comment]
Generally, in Python 3.x list() should be applied to dct.keys(), dct.items() or dct.values() only if the "dct" is changed during iteration. Otherwise it is not necessary and may be harmful - it makes code more verbose, less efficient and harder to read (one can expect `for k in list(dct.keys()):` body to modify the "dct").

**Comment by hleeldc** [comment]
The extra list() added done by 2to3.py. I wasn't paying too much attention to what it did.

**Comment by stevenbird** [comment]
Thanks for the feedback @kmike. I'm not so keen on creating our own local names for tkinter methods. The solution that @hleeldc contributed will not require compat and its extra code once we drop support for Python 2. So the problems you list will all go away.

**Comment by stevenbird** [comment]
Perhaps we should leave in the extra list() calls until we understand why 2to3.py added them.

**Pull request merged**